src/Chainweb/Pact/PactService.hs

```diff
@@ -418,41 +421,50 @@ finalizeCheckpointer finalize = do
 _liftCPErr :: Either String a -> PactServiceM cas a
 _liftCPErr = either internalError' return
 
+-- | The principal validation logic for groups of Pact Transactions.
+--
+-- Skips validation for genesis transactions, since gas accounts, etc. don't
+-- exist yet.
+--
 validateChainwebTxs
     :: PactDbEnv'
     -> Checkpointer
     -> BlockCreationTime
     -> BlockHeight
     -> Vector ChainwebTransaction
     -> IO (Vector Bool)
-validateChainwebTxs dbEnv cp blockOriginationTime bh txs =
-    V.mapM checkOne txs
+validateChainwebTxs dbEnv cp blockOriginationTime bh txs
+    | bh == 0 = pure $ V.replicate (V.length txs) False
+    | otherwise = V.mapM checkOne txs
   where
-    checkAccount validators tx = do
-      let !pm = P._pMeta $ payloadObj $ P._cmdPayload tx
-      let sender = P._pmSender pm
-          (P.GasLimit (P.ParsedInteger limit)) = P._pmGasLimit pm
-          (P.GasPrice (P.ParsedDecimal price)) = P._pmGasPrice pm
-          limitInCoin = price * fromIntegral limit
-      m <- readCoinAccount dbEnv sender
-      case m of
+    checkAccount :: [P.Command PayloadWithText -> Bool] -> P.Command PayloadWithText -> IO Bool
+    checkAccount validators tx = readCoinAccount dbEnv sender >>= \case
         Nothing -> return True
-        Just (T2 b _g) -> do
-          let validations = (const (b >= limitInCoin)) : validators
-          return $! all ($ tx) validations
+        Just (T2 b _) -> do
+            -- TODO Remove this one-off `limitInCoin` check, it will be subsumed
+            -- by the accumulated gas check.
+            let validations = gasCheck b : validators
+            return $! all ($ tx) validations
+      where
+        gasCheck :: Decimal -> b -> Bool
+        gasCheck b = const (b >= limitInCoin)
+
+        pm = P._pMeta . payloadObj $ P._cmdPayload tx
+        sender = P._pmSender pm
+        P.GasLimit (P.ParsedInteger limit) = P._pmGasLimit pm
+        P.GasPrice (P.ParsedDecimal price) = P._pmGasPrice pm
+        limitInCoin = price * fromIntegral limit
 
-    -- skip validation for genesis -- gas accounts / etc don't exist yet
-    checkOne tx = if bh == 0 then return True else checkOneReal tx
-    checkOneReal tx = do
+    checkOne :: P.Command PayloadWithText -> IO Bool
+    checkOne tx = do
```

Comment:
P.S. also I think this used to be/was supposed to be

`checkOne = if bh == 0 then const (return True) else checkOneReal`

which would only test bh once

Reply:
Once per Transaction, yes. Now it's once per call to `validateChainwebTxs`.